### PR TITLE
new implementation

### PR DESCRIPTION
**Pull request by Sgtfishtank**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -331,6 +342,9 @@ class HLSVod {
             }
 
             if (streamItem.get("subtitles")) {
+              if (!this.baseSubtitleUrl) {
+                reject(new Error("Missing default subtitle URL")); // Need a better error and name
```
Comment (Sgtfishtank) [inline index.js:346]: @birme I can't think of a better name for this variable and error message do you have a suggestion

Comment (Sgtfishtank) [reply index.js:346]: It was the splicer

File: index.js [hunk 2]
```diff
@@ -949,26 +1332,50 @@ class HLSVod {
...
+      if (subtitleGroupId) {
+        this._removeDoublediscontinuitiesFromExtraMedia(this.subtitleSegments)
+      }
+      if (this.shouldContainSubtitles) {
+        // we are doing all this to figure out the entire duration of the new vod so we can create a long subtitle segment that we can later chunk to smaller segments
+        let duration = this.getDuration();
+        const bandwidthsPrevVod = this.previousVod.getBandwidths();
+        const previousVodSeqCount = this.previousVod.getLiveMediaSequencesCount();
+        const lastMediaSequence = this.previousVod.getLiveMediaSequenceSegments(previousVodSeqCount - 1)[bandwidthsPrevVod[0]];
+        let tempdurr = 0;
```
Comment (birme) [inline index.js:1347]: Typo?

File: index.js [hunk 3]
```diff
@@ -910,6 +1223,65 @@ class HLSVod {
...
+              return false;
+            }
+          }
+          return true;
+        });
+      }
+    }
+  }
+
+  _generateSmallerSegments(segment, offset, leftover, useDummyUrl, first) {
```
Comment (birme) [inline index.js:1245]: Let's break this out as a non-class method and a standalone function. It makes it easier to unit-test the specific logic and test different edge cases

File: index.js [hunk 4]
```diff
@@ -910,6 +1223,65 @@ class HLSVod {
...
+        leftover = videoSegements[index].duration - totalSubtitleSegmentDuration
+        totalSubtitleSegmentDuration = 0;
+      }
+      newSegmentList.push(temp);
+      index++;
+    }
+    return { newSegments: newSegmentList, offset: index, leftover: leftover }
+  }
+
   _createMediaSequences() {
```
Comment (birme) [inline index.js:1285]: This function starts to grow. See if things can be split up into smaller functions

**Review by birme — CHANGES_REQUESTED** [review]

File: index.js [hunk 5]
```diff
@@ -4,6 +4,9 @@ const debug = require("debug")("hls-vodtolive");
 const verbose = require("debug")("hls-vodtolive-verbose");
 
 const { findIndexReversed, fetchWithRetry, urlResolve, segToM3u8 } = require("./utils.js");
+const { BADFAMILY } = require("dns");
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply index.js]: Nothing, removed.

File: index.js [hunk 6]
```diff
@@ -4,6 +4,9 @@ const debug = require("debug")("hls-vodtolive");
...
+const { URLSearchParams } = require("url");
+const { rename } = require("fs");
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply index.js]: Nothing, removed.

File: index.js [hunk 7]
```diff
@@ -4,6 +4,9 @@ const debug = require("debug")("hls-vodtolive");
 const verbose = require("debug")("hls-vodtolive-verbose");
 
 const { findIndexReversed, fetchWithRetry, urlResolve, segToM3u8 } = require("./utils.js");
+const { BADFAMILY } = require("dns");
+const { URLSearchParams } = require("url");
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply index.js]: OK will change to use nodejs.

File: index.js (hunk 2 above)
Comment (Sgtfishtank) [reply index.js]: Yes, renamed it

File: index.js [hunk 8]
```diff
@@ -3005,6 +3001,7 @@ class HLSVod {
       parser.on("m3u", (m3u) => {
         try {
           let initSegment = undefined;
+          let removedSegemntDuration = 0;
```
Comment (birme) [inline index.js:3004]: Please fix typo

File: index.js [hunk 9]
```diff
@@ -2876,7 +2872,7 @@ class HLSVod {
           if (m) {
             baseUrl = m[1] + "/";
           }
-
+          console.log(groupId, language, audioManifestUri, this.audioSegments)
```
Comment (birme) [inline index.js:2875]: Use debug() instead of console.log if you intend to keep this

**Review by birme — CHANGES_REQUESTED** [review]

**Comment by birme** [comment]
Don't know if this is about to be completed soon and ready to be merged to v3-release-candidate branch. I'd recommend to clean up the PR a bit first (use rebase to squash some commits) and rebase this branch with the current head of the v3-release-candidate branch

File: index.js (hunk 8 above)
Comment (Sgtfishtank) [reply index.js:3033]: fixed

File: index.js (hunk 9 above)
Comment (Sgtfishtank) [reply index.js]: this is removed, legacy from when testing

**Comment by Sgtfishtank** [comment]
> Don't know if this is about to be completed soon and ready to be merged to v3-release-candidate branch. I'd recommend to clean up the PR a bit first (use rebase to squash some commits) and rebase this branch with the current head of the v3-release-candidate branch

It should be completed soon, just making sure it is working with `eyevinn/channel-engin` first, and I will need to fix some of the previous comments first. Then I will rebase.

File: index.js (hunk 4 above)
Comment (Sgtfishtank) [reply index.js:1815]: it is now done

File: spec/hlsvod_audio_spec.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
+const HLSVod = require("../index.js");
+const fs = require("fs");
+const m3u8 = require("@eyevinn/m3u8");
+const exp = require("constants");
```
Comment (birme) [inline spec/hlsvod_audio_spec.js:4]: Remove?

Comment (Sgtfishtank) [reply spec/hlsvod_audio_spec.js]: yes unused

File: spec/hlsvod_audio_spec.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+const { subtle } = require("crypto");
+const Readable = require("stream").Readable;
+
+describe("HLSVod with subtitles", () => {
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply spec/hlsvod_audio_spec.js]: fixed

File: spec/hlsvod_audio_spec.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
+const HLSVod = require("../index.js");
+const fs = require("fs");
+const m3u8 = require("@eyevinn/m3u8");
+const exp = require("constants");
+const { subtle } = require("crypto");
```
[the remark this answers is not in the document]
Comment (Sgtfishtank) [reply spec/hlsvod_audio_spec.js]: correct, removed it

**Review by Nfrederiksen — APPROVED** [review]
